hw/ip/lc_ctrl/util/gen-lc-state-enc.py

```diff
@@ -0,0 +1,397 @@
+        exit(1)
+
+    if config['secded']['ecc_width'] != len(config['secded']['ecc_matrix']):
+        log.error('ECC matrix does not have correct number of rows')
+        exit(1)
+
+    for i, l in enumerate(config['secded']['ecc_matrix']):
+        for j, e in enumerate(l):
+            e = _check_int(e)
+            config['secded']['ecc_matrix'][i][j] = e
+
+
+def _is_valid_codeword(config, codeword):
+    '''Checks whether the bitstring is a valid ECC codeword.'''
+
+    data_width = config['secded']['data_width']
+    ecc_width = config['secded']['ecc_width']
+    if len(codeword) != (data_width + ecc_width):
+        log.error("Invalid codeword length {}".format(len(codeword)))
+        exit(1)
+
+    # Build syndrome and check whether it is zero.
+    syndrome = [0 for k in range(ecc_width)]
+
+    # The bitstring must be formatted as "data bits[N-1:0]" + "ecc bits[M-1:0]".
+    for j, fanin in enumerate(config['secded']['ecc_matrix']):
+        syndrome[j] = int(codeword[ecc_width - 1 - j])
+        for k in fanin:
+            syndrome[j] ^= int(codeword[ecc_width + data_width - 1 - k])
+
+    return sum(syndrome) == 0
+
+
+def _ecc_encode(config, dataword):
+    '''Calculate and prepend ECC bits.'''
+    if len(dataword) != config['secded']['data_width']:
+        log.error("Invalid codeword length {}".format(len(dataword)))
+        exit(1)
+
+    # Build syndrome
+    eccbits = ""
+    for j, fanin in enumerate(config['secded']['ecc_matrix']):
+        bit = 0
+        for k in fanin:
+            bit ^= int(dataword[config['secded']['data_width'] - 1 - k])
+        eccbits += format(bit, '01b')
+
+    return eccbits[::-1] + dataword
+
+
+def _is_incremental_codeword(word1, word2):
+    '''Test whether word2 is incremental wrt word1.'''
+    if len(word1) != len(word2):
+        log.error('Words are not of equal size')
+        exit(1)
+
+    _word1 = int(word1, 2)
+    _word2 = int(word2, 2)
+
+    # This basically checks that the second word does not
+    # clear any bits that are set to 1 in the first word.
+    return ((_word1 & _word2) == _word1)
+
+
+def _get_data_cand(mask, bits):
+    '''Scatter the bits into unset positions of mask.'''
+    j = 0
+    scatterword = ''
+    for b in mask:
+        if b == '1':
+            scatterword += '1'
+        else:
+            scatterword += bits[j]
+            j += 1
+
+    return scatterword
+
+
+def _get_hd(word1, word2):
+    '''Calculate Hamming distance between two words.'''
+    if len(word1) != len(word2):
+        log.error('Words are not of equal size')
+        exit(1)
+    return bin(int(word1, 2) ^ int(word2, 2)).count('1')
+
+
+def _get_incremental_codewords(config, base_ecc, existing_words):
+    '''Get all possible incremental codewords fulfilling the constraints.'''
+
+    base_data = base_ecc[config['secded']['ecc_width']:]
+
+    # Only need to spin through data bits that have not been set yet.
+    incr_cands = []
+    free_bits = base_data.count('0')
```

Comment:
i'm a little confused by this part.. if the base data were something like `001001`, wouldn't the range below end up being `1~2^4`?  

is that what you'd want..? sorry if i missed something.

Reply:
o wait, i see... i think what you're doing is find out what combination of those bits should be set, and then you scatter below, is that right? It would be great if you could leave a comment there for future readers like me :)

Reply:
yeah that's right. I first count how many bits are still set-able (i.e., the bits that are zero), and then just enumerate all possibilities for these bits. the enumeration bits are then scattered into the correct positions further below. I'll expand the comment here to explain this better.